controlplane/kubeadm/internal/control_plane.go

```diff
@@ -91,15 +95,35 @@ func NewControlPlane(ctx context.Context, managementCluster ManagementCluster, c
 		patchHelpers[machine.Name] = patchHelper
 	}
 
+	// Select machines that should be rolled out because of an outdated configuration or because rolloutAfter/Before expired.
+	reconciliationTime := metav1.Now()
```

Comment:
This change basically means that we already calculate all this when we create the ControlPlane vs calculating it when MachinesNeedingRollout & UpToDateMachines are called.

Does any of the data we use for the calculation change between the calls to NewControlPlane and MachinesNeedingRollout/UpToDateMachines?

(if yes we would end up with a change in behavior that we should look into)

In general this is absolutely a step in the right direction, we shouldn't recalculate this information multiple times in a single reconcile.

Reply:
Rollout decision is based on info from objects spec, and we are not changing objects spec during reconcile nor for KCP, nor for Machines (exception are the in place-propagated filelds, which are not relevant for rollout)

Reply:
Yeah I think the only thing that changes is the deletionTimestamp. But I guess tha was always after we called MachinesNeedingRollout/UpToDateMachines?

Reply:
> I think the only thing that changes is the deletionTimestamp. But I guess that was always after we called MachinesNeedingRollout/UpToDateMachines?

That's correct